src/pkg/utils/io.go

```diff
@@ -321,6 +322,127 @@ func SplitFile(path string, chunkSizeBytes int) (chunks [][]byte, sha256sum stri
+// the first file will be a metadata file containing:
+// - sha256sum of the original file
+// - number of bytes in the original file
+// - number of files the srcFile was split into
+// SplitFile will delete the original file
+//
+// Returns:
+// - fileNames: list of file paths srcFile was split across
+// - sha256sum: sha256sum of the srcFile before splitting
+// - err: any errors encountered
+func SplitFile(srcFile string, chunkSizeBytes int) (fileNames []string, sha256sum string, err error) {
+	hash := sha256.New()
+
+	// Set buffer size to some multiple of 4096 KiB for modern file system cluster sizes
+	bufferSize := 16 * 1024 * 1024 // 16 MiB
+	// if chunkSizeBytes is less than bufferSize, use chunkSizeBytes as bufferSize for simplicity
+	if chunkSizeBytes < bufferSize {
+		bufferSize = chunkSizeBytes
+	}
+	buf := make([]byte, bufferSize)
+
+	// get file size
+	fi, err := os.Stat(srcFile)
+	if err != nil {
+		return fileNames, "", err
+	}
+	fileSize := fi.Size()
+
+	// open file
+	file, err := os.Open(srcFile)
+	if err != nil {
+		return fileNames, "", err
+	}
+
+	// create file path starting from part 001
+	path := fmt.Sprintf("%s.part%03d", srcFile, len(fileNames)+1)
+	chunkFile, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return fileNames, "", err
+	}
+	fileNames = append(fileNames, path)
+	defer chunkFile.Close()
+
+	// setup counter for tracking how many bytes are left to write to file
+	chunkBytesRemaining := chunkSizeBytes
+	// Loop over the tarball hashing as we go and breaking it into chunks based on the chunkSizeBytes
+	for {
+		bytesRead, err := file.Read(buf)
+
+		if err != nil {
+			if err == io.EOF {
+				// At end of file, break out of loop
+				break
+			} else {
+				return fileNames, "", err
+			}
+		}
+
+		// Pass data to hash
+		hash.Write(buf[0:bytesRead])
+
+		// handle if we should split the data between two chunks
+		if chunkBytesRemaining < bytesRead {
+			// write the remaining chunk size to file
+			_, err := chunkFile.Write(buf[0:chunkBytesRemaining])
+			if err != nil {
+				return fileNames, "", err
+			}
+
+			// create new file
+			path = fmt.Sprintf("%s.part%03d", srcFile, len(fileNames)+1)
+			chunkFile, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY, 0644)
+			if err != nil {
+				return fileNames, "", err
+			}
+			fileNames = append(fileNames, path)
+			defer chunkFile.Close()
+
+			// write to new file where we left off
+			_, err = chunkFile.Write(buf[chunkBytesRemaining:bytesRead])
+			if err != nil {
+				return fileNames, "", err
+			}
+
+			// set chunkBytesRemaining considering how many bytes are already written to new file
+			chunkBytesRemaining = chunkSizeBytes - (bufferSize - chunkBytesRemaining)
+		} else {
+			_, err := chunkFile.Write(buf[0:bytesRead])
+			if err != nil {
+				return fileNames, "", err
+			}
+			chunkBytesRemaining = chunkBytesRemaining - bytesRead
+		}
+	}
+	file.Close()
```

Comment:
prob move to a `defer file.Close()` when `file` is initialized so that this function can't be leaky w/ file handles

Reply:
Do you know if you can remove an open file? Never tried before, but assumed it was required to close the file first.

Reply:
*nix systems can do it, but Windows has issues.

Only really a problem when you try to access that file again (like in this small change I PRed to `syft` https://github.com/anchore/syft/pull/1668).

Surprisingly static code analysis can't catch this, so I try to always `defer file.Close()` as soon as I open the file. _Then_ later if I want to explicitly close the file I can still do `file.Close` and just not handle the error (as you are doing now).

Reply:
Thanks! Makes sense. I've added that in